experimental/native_file_system/virtual_root_provider_win.cc

```diff
@@ -1,12 +1,63 @@
-// Copyright (c) 2014 Intel Corporation. All rights reserved.
-// Use of this source code is governed by a BSD-style license that can be
-// found in the LICENSE file.
-
-#include "xwalk/experimental/native_file_system/virtual_root_provider.h"
-
-#include "base/logging.h"
-
-VirtualRootProvider::VirtualRootProvider() {
-  // TODO(darktears): Windows support to be added.
-  NOTIMPLEMENTED();
-}
+// Copyright (c) 2015 Intel Corporation. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "xwalk/experimental/native_file_system/virtual_root_provider.h"
+
+#include <shlobj.h>
+
+#include "base/files/file_path.h"
+#include "base/files/file_util.h"
+
+bool VirtualRootProvider::testing_enabled_ = false;
+
+VirtualRootProvider::VirtualRootProvider() {
+  if (!virtual_root_map_.empty())
+    return;
+
+  if (testing_enabled_)
+    base::GetTempDir(&home_path_);
+  else
+    home_path_ = base::GetHomeDir();
+
+  LPWSTR wszPath = NULL;
+  HRESULT hr = SHGetKnownFolderPath(FOLDERID_Desktop,
+                                    KF_FLAG_DONT_UNEXPAND, NULL, &wszPath);
+  if (SUCCEEDED(hr))
+    virtual_root_map_["DESKTOP"] = base::FilePath(wszPath);
+
+  hr = SHGetKnownFolderPath(FOLDERID_Documents,
+                            KF_FLAG_DONT_UNEXPAND, NULL, &wszPath);
+  if (SUCCEEDED(hr))
+    virtual_root_map_["DOCUMENTS"] = base::FilePath(wszPath);
+
+  hr = SHGetKnownFolderPath(FOLDERID_Music,
+                            KF_FLAG_DONT_UNEXPAND, NULL, &wszPath);
+  if (SUCCEEDED(hr))
+    virtual_root_map_["MUSIC"] = base::FilePath(wszPath);
+
+  hr = SHGetKnownFolderPath(FOLDERID_Pictures,
+                            KF_FLAG_DONT_UNEXPAND, NULL, &wszPath);
+  if (SUCCEEDED(hr))
+    virtual_root_map_["PICTURES"] = base::FilePath(wszPath);
+
+  hr = SHGetKnownFolderPath(FOLDERID_Videos,
+                            KF_FLAG_DONT_UNEXPAND, NULL, &wszPath);
+  if (SUCCEEDED(hr))
+    virtual_root_map_["VIDEOS"] = base::FilePath(wszPath);
+
+  hr = SHGetKnownFolderPath(FOLDERID_Downloads,
+                            KF_FLAG_DONT_UNEXPAND, NULL, &wszPath);
+  if (SUCCEEDED(hr))
+    virtual_root_map_["DOWNLOADS"] = base::FilePath(wszPath);
+}
+
+void VirtualRootProvider::SetTesting(bool testing_enabled) {
+  testing_enabled_ = testing_enabled;
+  base::FilePath tmp_path;
+  base::GetTempDir(&tmp_path);
+  base::FilePath doc_path = tmp_path.Append(FILE_PATH_LITERAL("documents"));
+  if (!DirectoryExists(doc_path)) {
```

Comment:
I'd move this to the constructor (as all the other folders stuff is done there), and keep here only "testing_enabled_ = testing_enabled".